### ISSUE TITLE
pymco.exc.ConfigLookupError

### ISSUE THREAD

**Issue by saliceti** [rafaduran/python-mcollective#27]
We run mcollective installed with Puppet enterprise. We are using python 2.7.3 on Linux RedHat 6.4
mco works fine, but python-mcollective crashes, with the same configuration. When I try the ping example, I am getting a ConfigLookupError:
````
>>> from pymco import config,message,rpc
>>> config = config.Config.from_configfile('/var/lib/peadmin/.mcollective')
>>> msg = message.Message(body='ping', agent='discovery', config=config)
>>> rpc.SimpleAction(config=config, msg=msg, agent='discovery').call()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/rpc.py", line 61, in call
    self.connector.connect(wait=True)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/connector/__init__.py", line 58, in connect
    self.get_current_host_and_port())
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/config.py", line 172, in get_user_and_password
    current_host_and_port, connector))
pymco.exc.ConfigLookupError: ('127.0.0.1', 61613) is not in the configuration for activemq connector
````

The configuration file is the one generated by Puppet enterprise:
````
# This file managed by Puppet 3.6.2 (Puppet Enterprise 3.3.1)
main_collective = mcollective
collectives     = mcollective
libdir          = /opt/puppet/libexec/mcollective/
logfile         = /var/lib/peadmin/.mcollective.d/client.log
loglevel        = info

securityprovider = ssl
plugin.ssl_client_private = /var/lib/peadmin/.mcollective.d/peadmin-private.pem
plugin.ssl_client_public = /var/lib/peadmin/.mcollective.d/peadmin-public.pem
plugin.ssl_server_public = /var/lib/peadmin/.mcollective.d/mcollective-public.pem
plugin.ssl_serializer     = yaml

connector = activemq
plugin.activemq.pool.size = 1
plugin.activemq.pool.1.host = the_puppet_master_hostname
plugin.activemq.pool.1.port = 61613
plugin.activemq.pool.1.user = mcollective
plugin.activemq.pool.1.password = the_password
plugin.activemq.pool.1.ssl = true
plugin.activemq.pool.1.ssl.ca = /var/lib/peadmin/.mcollective.d/peadmin-cacert.pem
plugin.activemq.pool.1.ssl.key = /var/lib/peadmin/.mcollective.d/peadmin-private.pem
plugin.activemq.pool.1.ssl.cert = /var/lib/peadmin/.mcollective.d/peadmin-cert.pem

# Facts
factsource = yaml
plugin.yaml = /etc/puppetlabs/mcollective/facts.yaml
direct_addressing = 1
````

**Comment by saliceti** [comment]
@jantman  "the_puppet_master_hostname" is a placeholder for the the puppet master FQDN, where the activemq broker is. I'm not sure it's confidential, but I'd rather not take any risk. Same as "the_password"

I have tried your branch, same result.
What about "the additional logging"? Is there a log file?

Here is the traceback. Only the line numbers are different:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/rpc.py", line 65, in call
    self.connector.connect(wait=True)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/connector/__init__.py", line 61, in connect
    self.get_current_host_and_port())
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/config.py", line 178, in get_user_and_password
    current_host_and_port, connector))
pymco.exc.ConfigLookupError: ('127.0.0.1', 61613) is not in the configuration for activemq connector
```

**Comment by jantman** [comment]
@saliceti the logging uses the standard Python logging library. I.e., to get the maximum logging, put this between your imports and the rest of your code (where you work with pymco):

```
import logging
FORMAT = "[%(levelname)s %(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
logging.basicConfig(level=logging.DEBUG, format=FORMAT)
logger = logging.getLogger(__name__)
```

**Comment by saliceti** [comment]
OK interesting development! With the aditional logging, this is what I got:
````
[DEBUG config.py:56 -             __init__() ] initialized Config
[DEBUG rpc.py:20 -             __init__() ] init rpc.SimpleAction
[DEBUG rpc.py:64 -                 call() ] connecting, wait=True
[DEBUG config.py:108 -        get_connector() ] connector import path: pymco.connector.activemq.ActiveMQConnector
[DEBUG __init__.py:38 -             __init__() ] initializing connector
[INFO transport.py:637 -   attempt_connection() ] Attempting connection to host 127.0.0.1, port 61613
[INFO transport.py:670 -   attempt_connection() ] Established connection to host 127.0.0.1, port 61613
Traceback (most recent call last):
  File "test_pymco.py", line 9, in <module>
    rpc.SimpleAction(config=config, msg=msg, agent='discovery').call()
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/rpc.py", line 65, in call
    self.connector.connect(wait=True)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/connector/__init__.py", line 61, in connect
    self.get_current_host_and_port())
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/config.py", line 178, in get_user_and_password
    current_host_and_port, connector))
pymco.exc.ConfigLookupError: ('127.0.0.1', 61613) is not in the configuration for activemq connector
````
So we can see it tries to connect **before** the ConfigLookupError. Since I am running on the same server as the broker, the connection to 127.0.0.1 succeeds. That is why current_host_and_port is set to ('127.0.0.1', 61613), hence the ConfigLookupError.

To confirm, I have run the same code on another node and it does work (I haven't hidden the hostname for clarity):
````
/ubs/python-mcollective/test/lib/python2.7/site-packages/Crypto/Util/number.py:57: PowmInsecureWarning: Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.
  _warn("Not using mpz_powm_sec.  You should rebuild using libgmp >= 5 to avoid timing attack vulnerability.", PowmInsecureWarning)
[DEBUG config.py:56 -             __init__() ] initialized Config
[DEBUG rpc.py:20 -             __init__() ] init rpc.SimpleAction
[DEBUG rpc.py:64 -                 call() ] connecting, wait=True
[DEBUG config.py:108 -        get_connector() ] connector import path: pymco.connector.activemq.ActiveMQConnector
[DEBUG __init__.py:38 -             __init__() ] initializing connector
[INFO transport.py:637 -   attempt_connection() ] Attempting connection to host xldn0603nsk.ldn.swissbank.com, port 61613
[INFO transport.py:670 -   attempt_connection() ] Established connection to host xldn0603nsk.ldn.swissbank.com, port 61613
[INFO transport.py:479 -      __receiver_loop() ] Starting receiver loop
[DEBUG __init__.py:62 -              connect() ] connecting to middleware...
...
````
Please note **this only works with jantman's branch**. With the master I am getting this error after sending the frame:
````
[INFO transport.py:513 -      __receiver_loop() ] Receiver loop ended
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/sbcimp/run/pd/python/64-bit/2.7.3/lib/python2.7/threading.py", line 551, in __bootstrap_inner
    self.run()
  File "/sbcimp/run/pd/python/64-bit/2.7.3/lib/python2.7/threading.py", line 504, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/stomp/transport.py", line 491, in __receiver_loop
    self.process_frame(f, frame)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/stomp/transport.py", line 407, in process_frame
    self.notify(frame_type, f.headers, f.body)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/stomp/transport.py", line 469, in notify
    rtn = notify_func(headers, body)
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/listener.py", line 85, in on_message
    self.responses.append(self.security.decode(body))
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/security/__init__.py", line 64, in decode
    return self.verify(self.deserialize(msg))
  File "/ubs/python-mcollective/test/lib/python2.7/site-packages/python_mcollective-0.0.1.dev3-py2.7.egg/pymco/security/ssl.py", line 44, in verify
    hash_ = SHA.new(msg[':body'].encode('utf8'))
KeyError: ':body'
````

**Comment by jantman** [comment]
The KeyError is what was discussed here: https://github.com/rafaduran/python-mcollective/pull/25#commitcomment-9099267

Also... if I'm remembering things correctly... instead of connecting to '127.0.0.1', try connecting to the machine's actual hostname, i.e. the one that the SSL certificate is generated for, and that has valid reverse DNS. I just remembered this, but I'm pretty sure mco requires that.

**Comment by rafaduran** [comment]
@jantman @saliceti I think the pymco.exc.ConfigLookupError error is due to how pymco creates STOMP connection, since stomp.py by default tries to connect using loopback interface and there is no entry in the configuration file for it. Setting ``try_loopback_connect`` to False should solve the issue. The other one as @jantman points should be fixed once #25 is merged.
